### ISSUE TITLE
Strange URL (http://<nil>) on public endpoints

### ISSUE THREAD

**Issue by MiTschMR** [rancher/rancher#17876]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
1. Set up Rancher HA as described in [https://rancher.com/docs/rancher/v2.x/en/installation/ha/](url)
2. Deploy a Catalog App which creates a public endpoint

**Result:**
The public endpoints have strange URLs, for example: 
- http://\<nil>:32285
- http://\<nil>:31871

**Other details that may be helpful:**
The issue does not occur when using single-node Installation of Rancher. 

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): **Rancher v2.1.5, but even tried with v2.1.6**
- Installation option (single install/HA): **HA**

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): **Imported**
- Machine type (cloud/VM/metal) and specifications (CPU/memory): **4 VMs, each with 2 vCPUs and 16 GB RAM. Two hold the etcd and Control Plane roles, the other two are workers.**

- cluster.yml (used for creation): 
```
# If you intened to deploy Kubernetes in an air-gapped environment,
# please consult the documentation on how to configure custom RKE images.
nodes:
- address: dkr9000.blab.ch
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - etcd
  hostname_override: ""
  user: user
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: /home/user/.ssh/id_rsa
  labels: {}
- address: dkr9001.blab.ch
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - etcd
  hostname_override: ""
  user: user
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: /home/user/.ssh/id_rsa
  labels: {}
- address: dkr9002.blab.ch
  port: "22"
  internal_address: ""
  role:
  - worker
  hostname_override: ""
  user: user
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: /home/user/.ssh/id_rsa
  labels: {}
- address: dkr9003.blab.ch
  port: "22"
  internal_address: ""
  role:
  - worker
  hostname_override: ""
  user: user
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: /home/user/.ssh/id_rsa
  labels: {}
services:
  etcd:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    external_urls: []
    ca_cert: ""
    cert: ""
    key: ""
    path: ""
    snapshot: null
    retention: ""
    creation: ""
  kube-api:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    service_cluster_ip_range: 10.43.0.0/16
    service_node_port_range: ""
    pod_security_policy: false
  kube-controller:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    cluster_cidr: 10.42.0.0/16
    service_cluster_ip_range: 10.43.0.0/16
  scheduler:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
  kubelet:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    cluster_domain: cluster.local
    infra_container_image: ""
    cluster_dns_server: 10.43.0.10
    fail_swap_on: false
  kubeproxy:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
network:
  plugin: canal
  options: {}
authentication:
  strategy: x509
  options: {}
  sans: []
addons: ""
addons_include: []
system_images:
  etcd: rancher/coreos-etcd:v3.2.18
  alpine: rancher/rke-tools:v0.1.15
  nginx_proxy: rancher/rke-tools:v0.1.15
  cert_downloader: rancher/rke-tools:v0.1.15
  kubernetes_services_sidecar: rancher/rke-tools:v0.1.15
  kubedns: rancher/k8s-dns-kube-dns-amd64:1.14.10
  dnsmasq: rancher/k8s-dns-dnsmasq-nanny-amd64:1.14.10
  kubedns_sidecar: rancher/k8s-dns-sidecar-amd64:1.14.10
  kubedns_autoscaler: rancher/cluster-proportional-autoscaler-amd64:1.0.0
  kubernetes: rancher/hyperkube:v1.11.6-rancher1
  flannel: rancher/coreos-flannel:v0.10.0
  flannel_cni: rancher/coreos-flannel-cni:v0.3.0
  calico_node: rancher/calico-node:v3.1.3
  calico_cni: rancher/calico-cni:v3.1.3
  calico_controllers: ""
  calico_ctl: rancher/calico-ctl:v2.0.0
  canal_node: rancher/calico-node:v3.1.3
  canal_cni: rancher/calico-cni:v3.1.3
  canal_flannel: rancher/coreos-flannel:v0.10.0
  wave_node: weaveworks/weave-kube:2.1.2
  weave_cni: weaveworks/weave-npc:2.1.2
  pod_infra_container: rancher/pause-amd64:3.1
  ingress: rancher/nginx-ingress-controller:0.16.2-rancher1
  ingress_backend: rancher/nginx-ingress-controller-defaultbackend:1.4
  metrics_server: rancher/metrics-server-amd64:v0.2.1
ssh_key_path: /home/user/.ssh/id_rsa
ssh_agent_auth: true
authorization:
  mode: rbac
  options: {}
ignore_docker_version: false
kubernetes_version: ""
private_registries: []
ingress:
  provider: ""
  options: {}
  node_selector: {}
  extra_args: {}
cluster_name: "test"
cloud_provider:
  name: ""
prefix_path: ""
addon_job_timeout: 0
bastion_host:
  address: ""
  port: ""
  user: ""
  ssh_key: ""
  ssh_key_path: ""
monitoring:
  provider: ""
  options: {}
```

- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.2", GitCommit:"cff46ab41ff0bb44d8584413b598ad8360ec1def", GitTreeState:"clean", BuildDate:"2019-01-10T23:35:51Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.6", GitCommit:"b1d75deca493a24a2f87eb1efde1a569e52fc8d9", GitTreeState:"clean", BuildDate:"2018-12-16T04:30:10Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
Client:
 Version:      17.03.2-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   f5ec1e2
 Built:        Tue Jun 27 02:21:36 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.2-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   f5ec1e2
 Built:        Tue Jun 27 02:21:36 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

- Operating System is CentOS 7: 
 **Output of cat /etc/os-release:**
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

 **Output of uname -r:**
```
3.10.0-957.1.3.el7.x86_64
```

**Comment by MiTschMR** [comment]
Upgrade to Rancher v2.1.6 and Kubernetes v1.12.5 did not resolve the issue.

**Comment by MiTschMR** [comment]
Example of what I exactly mean: 

**Software:** 
- OS: CentOS 7
- Rancher version: v2.1.6
- Kubernetes version: v1.12.4
- Docker version: 17.3.2
- RKE version: v0.1.15

**HA-Setup:** 
- VM 1 is loadbalancer
-  VM 2 and 3 are etcd/Control Plane hosts
- VM 4 and 5 are workers

Connection can be established, it is all working correctly (no errors). 

To reproduce the problem, I use "efk" from the "Library"-Catalog. I go to Cluster --> test --> Namespace Default. Then I go to Catalog Apps, select EFK and change the setting "Expose Kibana using Layer 7 Load Balancer" to "False". After that, I launch the application. The following occurs: 
- The Pods, Services, Secrets and Config Maps are created
- The Endpoints are created

Now, let's have a look at the Endpoints: 
![app-efk-test](https://user-images.githubusercontent.com/47354974/52479012-7856d480-2ba7-11e9-88cc-e4a101ceaf17.PNG)

The URL is http://\<nil>:Port

Here the output of `kubectl get service -n efk-test`: 
```
NAME                     TYPE       CLUSTER-IP      EXTERNAL-IP   PORT(S)                         AGE
efk-test-elasticsearch   NodePort   10.43.184.136   <none>        9200:31237/TCP,9300:31986/TCP   5m37s
efk-test-kibana          NodePort   10.43.72.147    <none>        443:31290/TCP                   5m37s
```

Output of `kubectl describe service -n efk-test efk-test-elasticsearch`:
```
Name:                     efk-test-elasticsearch
Namespace:                efk-test
Labels:                   app=elasticsearch
                          chart=elasticsearch-0.1.0
                          heritage=Tiller
                          io.cattle.field/appId=efk-test
                          release=efk-test
Annotations:              field.cattle.io/publicEndpoints:
                            [{"addresses":["\u003cnil\u003e"],"port":31237,"protocol":"TCP","serviceName":"efk-test:efk-test-elasticsearch","allNodes":true},{"address...
Selector:                 app=elasticsearch,component=master,release=efk-test
Type:                     NodePort
IP:                       10.43.184.136
Port:                     http  9200/TCP
TargetPort:               http/TCP
NodePort:                 http  31237/TCP
Endpoints:                10.42.2.27:9200,10.42.2.28:9200,10.42.2.29:9200 + 3 more...
Port:                     transport  9300/TCP
TargetPort:               transport/TCP
NodePort:                 transport  31986/TCP
Endpoints:                10.42.2.27:9300,10.42.2.28:9300,10.42.2.29:9300 + 3 more...
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```

Output of `kubectl describe service -n efk-test efk-test-elasticsearch`:
```
Name:                     efk-test-kibana
Namespace:                efk-test
Labels:                   app=kibana
                          chart=kibana-0.5.0
                          heritage=Tiller
                          io.cattle.field/appId=efk-test
                          release=efk-test
Annotations:              field.cattle.io/publicEndpoints:
                            [{"addresses":["\u003cnil\u003e"],"port":31290,"protocol":"TCP","serviceName":"efk-test:efk-test-kibana","allNodes":true}]
Selector:                 app=kibana,release=efk-test
Type:                     NodePort
IP:                       10.43.72.147
Port:                     <unset>  443/TCP
TargetPort:               5601/TCP
NodePort:                 <unset>  31290/TCP
Endpoints:                10.42.3.39:5601
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```

As you can see in the Annotations, the adress does not look healthy 😄

**Comment by vincent99** [comment]
FYI your "HA" is not HA at all... 2 nodes with etcd is _worse_ than one, it survives a maximum of 0 nodes failing and you now have twice as many available to fail.  You want an odd number.

**Comment by MiTschMR** [comment]
I admit this is my fault, I will change that soon. But unfortunately this does not solve my problem. 

If you need more information, configurations or logs, please tell me.

**Comment by vincent99** [comment]
Yes I know it's not related, just saying.

**Comment by MiTschMR** [comment]
Hello everyone

I have now tried it again with two HP ProLiant BL490c G7 and three HP ProLiant BL620c G7. The problem is still the same when deploying workloads with endpoints: 

http://<nil>:32285
http://<nil>:31871 or other URLs like this. 

This time I have used Rancher version v2.2.3 and RKE version v0.2.2. 

Other details that may be helpful:
The issue does not occur when using single-node Installation of Rancher.

**Comment by lllamnyp** [comment]
I am experiencing this too. What's strange is that the first cluster I brought up doesn't have this problem. The next two, however, do. It would be helpful to know, where rancher gets the info to populate the annotations field
```
field.cattle.io/publicEndpoints:
                            [{"addresses":["\u003cnil\u003e"]...
```

**Comment by axeal** [comment]
From my testing it appears that this behaviour occurs when provisioning clusters with nodes where the 'address' defined for cluster nodes in the RKE cluster configuration yaml is a FQDN, rather than an IP, I reproduced with the cluster configuration yaml below, using RKE v0.2.7 and Rancher v2.2.7:

```
nodes:
- address: node00.alexseymour.com
  user: root
  role: [controlplane,worker,etcd]
- address: node01.alexseymour.com
  user: root
  role: [controlplane,worker,etcd]
- address: node02.alexseymour.com
  user: root
  role: [controlplane,worker,etcd]

ssh_agent_auth: true

services:
  etcd:
    snapshot: true
    creation: 6h
    retention: 24h
```

**Comment by alena1108** [comment]
@axeal can you share the following output for one of the nodes having fqdn as the name:

`kubectl get node <nodename> -o json`

**Comment by lllamnyp** [comment]
@axeal I can confirm your observation. I was getting this quite consistently when using the fqdn of our servers with rke0.2.4 and rancher 2.2.4.

**Comment by axeal** [comment]
@alena1108 
```bash
$ kubectl get node node00.alexseymour.com -o json
{
    "apiVersion": "v1",
    "kind": "Node",
    "metadata": {
        "annotations": {
            "flannel.alpha.coreos.com/backend-data": "{\"VtepMAC\":\"e6:d2:2c:12:0e:5a\"}",
            "flannel.alpha.coreos.com/backend-type": "vxlan",
            "flannel.alpha.coreos.com/kube-subnet-manager": "true",
            "flannel.alpha.coreos.com/public-ip": "167.71.52.36",
            "node.alpha.kubernetes.io/ttl": "0",
            "rke.cattle.io/external-ip": "node00.alexseymour.com",
            "rke.cattle.io/internal-ip": "node00.alexseymour.com",
            "volumes.kubernetes.io/controller-managed-attach-detach": "true"
        },
        "creationTimestamp": "2019-09-23T07:47:19Z",
        "labels": {
            "beta.kubernetes.io/arch": "amd64",
            "beta.kubernetes.io/os": "linux",
            "kubernetes.io/arch": "amd64",
            "kubernetes.io/hostname": "node00.alexseymour.com",
            "kubernetes.io/os": "linux",
            "node-role.kubernetes.io/controlplane": "true",
            "node-role.kubernetes.io/etcd": "true",
            "node-role.kubernetes.io/worker": "true"
        },
        "name": "node00.alexseymour.com",
        "resourceVersion": "2711",
        "selfLink": "/api/v1/nodes/node00.alexseymour.com",
        "uid": "5ef04b4a-ddd6-11e9-8634-bede9c943c2a"
    },
    "spec": {
        "podCIDR": "10.42.0.0/24"
    },
    "status": {
        "addresses": [
            {
                "address": "167.71.52.36",
                "type": "InternalIP"
            },
            {
                "address": "node00.alexseymour.com",
                "type": "Hostname"
            }
        ],
        "allocatable": {
            "cpu": "2",
            "ephemeral-storage": "74761043435",
            "hugepages-1Gi": "0",
            "hugepages-2Mi": "0",
            "memory": "3943916Ki",
            "pods": "110"
        },
        "capacity": {
            "cpu": "2",
            "ephemeral-storage": "81120924Ki",
            "hugepages-1Gi": "0",
            "hugepages-2Mi": "0",
            "memory": "4046316Ki",
            "pods": "110"
        },
        "conditions": [
            {
                "lastHeartbeatTime": "2019-09-23T07:50:19Z",
                "lastTransitionTime": "2019-09-23T07:47:19Z",
                "message": "kubelet has sufficient memory available",
                "reason": "KubeletHasSufficientMemory",
                "status": "False",
                "type": "MemoryPressure"
            },
            {
                "lastHeartbeatTime": "2019-09-23T07:50:19Z",
                "lastTransitionTime": "2019-09-23T07:47:19Z",
                "message": "kubelet has no disk pressure",
                "reason": "KubeletHasNoDiskPressure",
                "status": "False",
                "type": "DiskPressure"
            },
            {
                "lastHeartbeatTime": "2019-09-23T07:50:19Z",
                "lastTransitionTime": "2019-09-23T07:47:19Z",
                "message": "kubelet has sufficient PID available",
                "reason": "KubeletHasSufficientPID",
                "status": "False",
                "type": "PIDPressure"
            },
            {
                "lastHeartbeatTime": "2019-09-23T07:50:19Z",
                "lastTransitionTime": "2019-09-23T07:47:49Z",
                "message": "kubelet is posting ready status",
                "reason": "KubeletReady",
                "status": "True",
                "type": "Ready"
            }
        ],
        "daemonEndpoints": {
            "kubeletEndpoint": {
                "Port": 10250
            }
        },
        "images": [
            {
                "names": [
                    "rancher/hyperkube@sha256:0fc3e30f679ed144cb99262fea0357334844e4eb14f673a4257026115eba81a2",
                    "rancher/hyperkube:v1.14.5-rancher1"
                ],
                "sizeBytes": 1051241871
            },
            {
                "names": [
                    "rancher/nginx-ingress-controller@sha256:5a60a68777ed1aa5472b987d1f085a653ada50056b7e6a43d22361973a24f2c0",
                    "rancher/nginx-ingress-controller:0.21.0-rancher3"
                ],
                "sizeBytes": 568544147
            },
            {
                "names": [
                    "rancher/rancher@sha256:2351fcc560585f5d236ef5f758a2ad6dafa95279381dcac01b91455ee6a6ea8b",
                    "rancher/rancher:v2.2.7"
                ],
                "sizeBytes": 473735479
            },
            {
                "names": [
                    "rancher/rancher-agent@sha256:fe9f7981b7ef1f585ec1ad7cb0dcc90da0f8c81d098f15abfca2bbfcda492213",
                    "rancher/rancher-agent:v2.2.7"
                ],
                "sizeBytes": 280450684
            },
            {
                "names": [
                    "rancher/rke-tools@sha256:d48bf7d13461694a5d2a8aefea48b9276368e67431e935420d1565844d495077",
                    "rancher/rke-tools:v0.1.40"
                ],
                "sizeBytes": 130122333
            },
            {
                "names": [
                    "rancher/calico-node@sha256:2f46157483904649f334da83e6b5e522bf20cbd3ea6ec9f43456673933021d5a",
                    "rancher/calico-node:v3.4.0"
                ],
                "sizeBytes": 75919845
            },
            {
                "names": [
                    "rancher/calico-cni@sha256:b3f4045680586133039f1e77f3ebb4bb5e6c60fa85f184f395eee32303462715",
                    "rancher/calico-cni:v3.4.0"
                ],
                "sizeBytes": 75434816
            },
            {
                "names": [
                    "quay.io/jetstack/cert-manager-controller@sha256:958074178ed595cd6733ecf07b746e520bddae10c6264c70cb8b07ecb0083311",
                    "quay.io/jetstack/cert-manager-controller:v0.5.2"
                ],
                "sizeBytes": 47280337
            },
            {
                "names": [
                    "rancher/coreos-flannel@sha256:57f85f0bb7360f2dd378d74ffc01ac99a4c152ad463455ff429867c6879f4386",
                    "rancher/coreos-flannel:v0.10.0"
                ],
                "sizeBytes": 44598861
            },
            {
                "names": [
                    "rancher/coredns-coredns@sha256:638adb0319813f2479ba3642bbe37136db8cf363b48fb3eb7dc8db634d8d5a5b",
                    "rancher/coredns-coredns:1.3.1"
                ],
                "sizeBytes": 40303560
            },
            {
                "names": [
                    "rancher/coreos-etcd@sha256:b2100f7b0dbd592e6e81689ac607055ed80f0bdbde7e29d24b4decf4cca63ea9",
                    "rancher/coreos-etcd:v3.3.10-rancher1"
                ],
                "sizeBytes": 39468433
            },
            {
                "names": [
                    "rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404d660b32517b18fcbf0144",
                    "rancher/pause:3.1"
                ],
                "sizeBytes": 742472
            }
        ],
        "nodeInfo": {
            "architecture": "amd64",
            "bootID": "85f123c6-707d-4474-99b5-c3ab7b00fe45",
            "containerRuntimeVersion": "docker://18.9.9",
            "kernelVersion": "4.4.0-164-generic",
            "kubeProxyVersion": "v1.14.5",
            "kubeletVersion": "v1.14.5",
            "machineID": "311936e2303b034fe7ef70182235b8cb",
            "operatingSystem": "linux",
            "osImage": "Ubuntu 16.04.6 LTS",
            "systemUUID": "62368578-A16B-4AE0-BE9E-5952516FAB2D"
        }
    }
}
```

**Comment by alena1108** [comment]
This issue has been fixed in v2.3.0 with https://github.com/rancher/rancher/pull/22287; corresponding issue https://github.com/rancher/rancher/issues/19208.

Closing as a duplicate.
